smart_open/smart_open_lib.py

```diff
@@ -203,25 +220,16 @@ class S3OpenRead(object):
     Implement streamed reader from S3, as an iterable & context manager.
 
     """
-    def __init__(self, parsed_uri):
-        if parsed_uri.scheme not in ("s3", "s3n"):
-            raise TypeError("can only process S3 files")
-        self.parsed_uri = parsed_uri
-        s3_connection = boto.connect_s3(
-            aws_access_key_id=parsed_uri.access_id,
-            aws_secret_access_key=parsed_uri.access_secret)
-        self.read_key = s3_connection.get_bucket(parsed_uri.bucket_id).lookup(parsed_uri.key_id)
-        if self.read_key is None:
-            raise KeyError(parsed_uri.key_id)
+    def __init__(self, read_key):
+        if not isinstance(read_key, boto.s3.key.Key):
```

Comment:
I added this because `S3OpenRead` is a class that could be used directly by users of the library, so I don't think we can assume the type checks in `smart_open` will have been performed already. Isn't there a way to check if this was an instance of `boto.s3.key.Key` or the corresponding mock library class?